File: app/src/main/java/sgtmelon/scriptum/element/common/OptionsDialog.java
package sgtmelon.scriptum.element.common;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;
import sgtmelon.scriptum.office.annot.def.DialogDef;

public class OptionsDialog extends DialogFragment implements DialogInterface.OnClickListener {

    private Context context;

    private String[] items;
    private int position;

    private DialogInterface.OnClickListener onClickListener;

    public void setArguments(String[] items, int p) {
        Bundle arg = new Bundle();

        arg.putStringArray(DialogDef.INIT, items);
        arg.putInt(DialogDef.VALUE, p);

        setArguments(arg);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        this.context = context;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        Bundle arg = getArguments();

        if (savedInstanceState != null) {
            items = savedInstanceState.getStringArray(DialogDef.INIT);
            position = savedInstanceState.getInt(DialogDef.VALUE);
        } else if (arg != null) {
            items = arg.getStringArray(DialogDef.INIT);
            position = arg.getInt(DialogDef.VALUE);
        }

        return new AlertDialog.Builder(context)
                .setItems(items, this)
                .setCancelable(true)
                .create();
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);

        outState.putStringArray(DialogDef.INIT, items);
        outState.putInt(DialogDef.VALUE, position);
    }

    public void setOnClickListener(DialogInterface.OnClickListener onClickListener) {
        this.onClickListener = onClickListener;
    }

    public int getPosition() {
        return position;
    }

    @Override
    public void onClick(DialogInterface dialogInterface, int i) {
        if (onClickListener != null) onClickListener.onClick(dialogInterface, i);
    }

}

File: safedialog/src/main/java/sgtmelon/safedialog/OptionsDialog.kt
package sgtmelon.safedialog

import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import sgtmelon.safedialog.annotation.NdValue
import sgtmelon.safedialog.annotation.SavedTag

/**
 * Dialog showing options with list item
 */
class OptionsDialog : DialogFragment(), DialogInterface.OnClickListener {

    var title: String = NdValue.TEXT

    var itemListener: DialogInterface.OnClickListener? = null
    var dismissListener: DialogInterface.OnDismissListener? = null

    private var itemList: List<String> = ArrayList()

    /**
     * Save item position in list for next operations
     */
    var position: Int = NdValue.POSITION
        private set

    /**
     * Call before [safeShow].
     */
    fun setArguments(itemArray: Array<String>, p: Int) = apply {
        arguments = Bundle().apply {
            putStringArray(SavedTag.INIT, itemArray)
            putInt(SavedTag.VALUE, p)
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        if (savedInstanceState != null) {
            onRestoreContentState(savedInstanceState)
        }

        onRestoreInstanceState(bundle = savedInstanceState ?: arguments)

        val builder = AlertDialog.Builder(context as Context)
            .setItems(itemList.toTypedArray(), this)
            .setCancelable(true)

        if (title != NdValue.TEXT) {
            builder.setTitle(title)
        }

        return builder.create().applyAnimation()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString(SavedTag.TITLE, title)
        outState.putStringArray(SavedTag.INIT, itemList.toTypedArray())
        outState.putInt(SavedTag.VALUE, position)
    }

    /**
     * Use for restore dialog content which was written before [safeShow]
     * (e.g. title, nameArray and ect.).
     *
     * Call inside [onCreateDialog] before create them.
     */
    private fun onRestoreContentState(savedInstanceState: Bundle) {
        title = savedInstanceState.getString(SavedTag.TITLE) ?: NdValue.TEXT
    }

    /**
     * Function for restore content which was passed throw setArgument functions
     * (e.g. position, check and ect.).
     */
    private fun onRestoreInstanceState(bundle: Bundle?) {
        itemList = bundle?.getStringArray(SavedTag.INIT)?.toList() ?: ArrayList()
        position = bundle?.getInt(SavedTag.VALUE) ?: NdValue.POSITION
    }

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)
        dismissListener?.onDismiss(dialog)
    }

    override fun onClick(dialogInterface: DialogInterface, i: Int) {
        itemListener?.onClick(dialogInterface, i)
    }
}